File: databricks_app/services/vendor_list_ai.py
import json
import logging
from datetime import datetime, timezone
from typing import Optional

from .llm import llm_complete, is_provider_available, get_available_providers, PROVIDER_LABELS

logger = logging.getLogger(__name__)


def _build_equipment_data_string(equipment: list[dict]) -> str:
    lines = []
    for i, eq in enumerate(equipment):
        parts = [
            f"Equipment {i + 1}:",
            f"  ID: {eq.get('id', '')}",
            f"  Type: {eq.get('equipmentType', eq.get('equipment_type', ''))}",
            f"  Process: {eq.get('process', '')}",
            f"  Quantity: {eq.get('quantity', 1)}",
            f"  Description: {eq.get('description', '')}",
            f"  Design Basis: {eq.get('designBasis', eq.get('design_basis', ''))}",
            "  Specs:",
        ]
        specs = eq.get("specs", {})
        if isinstance(specs, dict):
            for key, spec in specs.items():
                if isinstance(spec, dict):
                    val = spec.get("value", "")
                    unit = spec.get("unit", "")
                    parts.append(f"    {key}: {val} {unit}".rstrip())
                else:
                    parts.append(f"    {key}: {spec}")
        notes = eq.get("notes", "")
        if notes:
            parts.append(f"  Notes: {notes}")
        lines.append("\n".join(parts))
    return "\n\n".join(lines)


async def _get_prompt_template(key: str, storage=None) -> str:
    if storage and hasattr(storage, "get_prompt_template_by_key"):
        try:
            db_template = storage.get_prompt_template_by_key(key)
            if db_template and db_template.get("template"):
                return db_template["template"]
        except Exception:
            pass
    from ..knowledge_base.default_prompts import DEFAULT_PROMPTS
    prompt = DEFAULT_PROMPTS.get(key)
    if prompt:
        return prompt.get("template", "")
    return ""


async def generate_vendor_list_with_ai(
    equipment: list[dict],
    project_type: str,
    preferred_model: str,
    storage=None,
) -> dict:
    template = await _get_prompt_template("vendor_list", storage)
    if not template:
        raise ValueError("Vendor list prompt template not found")

    equipment_data = _build_equipment_data_string(equipment)
    project_context = f"Project Type: {project_type}"

    prompt = template.replace("{{EQUIPMENT_DATA}}", equipment_data)
    prompt = prompt.replace("{{PROJECT_CONTEXT}}", project_context)

    providers_to_try = []
    if is_provider_available(preferred_model):
        providers_to_try.append(preferred_model)
    for p in get_available_providers():
        if p not in providers_to_try:
            providers_to_try.append(p)

    if not providers_to_try:
        raise RuntimeError("No LLM providers available for vendor list generation")

    last_error = None
    for provider in providers_to_try:
        try:
            logger.info("Vendor List: Trying provider %s...", provider)
            result = llm_complete(
                model=provider,
                messages=[
                    {"role": "system", "content": prompt},
                    {"role": "user", "content": "Generate the recommended vendor list for the equipment listed above. Return only valid JSON."},
                ],
                max_tokens=8192,
                json_mode=True,
            )

            content = result["content"].strip()
            import re
            fence_match = re.search(r"```(?:json)?\s*([\s\S]*?)```", content)
            if fence_match:
                content = fence_match.group(1).strip()

            try:
                parsed = json.loads(content)
            except json.JSONDecodeError as e:
                raise ValueError(f"Failed to parse vendor list JSON: {e}")

            items = parsed.get("items")
            if not isinstance(items, list):
                raise ValueError("Vendor list response missing 'items' array")

            for item in items:
                if not item.get("equipmentId") and not item.get("equipment_id"):
                    raise ValueError("Vendor list item missing equipmentId")
                if "equipment_id" in item and "equipmentId" not in item:
                    item["equipmentId"] = item.pop("equipment_id")
                if "equipment_type" in item and "equipmentType" not in item:
                    item["equipmentType"] = item.pop("equipment_type")
                if "specs_summary" in item and "specsSummary" not in item:
                    item["specsSummary"] = item.pop("specs_summary")
                item.setdefault("quantity", 1)
                item.setdefault("specsSummary", "")
                item.setdefault("process", "")
                if not isinstance(item.get("recommendations"), list):
                    item["recommendations"] = []
                for rec in item["recommendations"]:
                    if "spec_sheet_url" in rec and "specSheetUrl" not in rec:
                        rec["specSheetUrl"] = rec.pop("spec_sheet_url")
                    if "website_url" in rec and "websiteUrl" not in rec:
                        rec["websiteUrl"] = rec.pop("website_url")
                    if "model_number" in rec and "modelNumber" not in rec:
                        rec["modelNumber"] = rec.pop("model_number")

            provider_label = PROVIDER_LABELS.get(provider, provider)
            vendor_list = {
                "items": items,
                "generatedAt": datetime.now(timezone.utc).isoformat(),
                "modelUsed": provider_label,
            }

            return {"vendor_list": vendor_list, "provider_label": provider_label}

        except Exception as e:
            last_error = e
            logger.warning("Vendor List: Provider %s failed: %s", provider, str(e))

    raise last_error or RuntimeError("All providers failed for vendor list generation")

File: server/services/vendorListAI.ts
import { llmComplete, isProviderAvailable, getAvailableProviders, providerLabels, type LLMProvider } from "../llm";
import type { EquipmentItem, VendorList } from "@shared/schema";
import type { PromptKey } from "@shared/default-prompts";
import { DEFAULT_PROMPTS } from "@shared/default-prompts";

function buildEquipmentDataString(equipment: EquipmentItem[]): string {
  return equipment.map((eq, i) => {
    const specLines = Object.entries(eq.specs)
      .map(([key, spec]) => {
        const val = (spec.value !== null && typeof spec.value === "object" && "value" in spec.value)
          ? String((spec.value as any).value)
          : String(spec.value ?? "");
        return `    ${key}: ${val} ${spec.unit || ""}`.trim();
      })
      .join("\n");

    return [
      `Equipment ${i + 1}:`,
      `  ID: ${eq.id}`,
      `  Type: ${eq.equipmentType}`,
      `  Process: ${eq.process}`,
      `  Quantity: ${eq.quantity}`,
      `  Description: ${eq.description}`,
      `  Design Basis: ${eq.designBasis}`,
      `  Specs:`,
      specLines,
      eq.notes ? `  Notes: ${eq.notes}` : "",
    ].filter(Boolean).join("\n");
  }).join("\n\n");
}

async function getPromptTemplate(key: PromptKey, storage?: any): Promise<string> {
  if (storage && typeof storage.getPromptTemplateByKey === "function") {
    try {
      const dbTemplate = await storage.getPromptTemplateByKey(key);
      if (dbTemplate?.template) return dbTemplate.template;
    } catch {
    }
  }
  return DEFAULT_PROMPTS[key]?.template || "";
}

export async function generateVendorListWithAI(
  equipment: EquipmentItem[],
  projectType: string,
  preferredModel: LLMProvider,
  storage?: any,
): Promise<{ vendorList: VendorList; providerLabel: string }> {
  const template = await getPromptTemplate("vendor_list", storage);
  if (!template) {
    throw new Error("Vendor list prompt template not found");
  }

  const equipmentData = buildEquipmentDataString(equipment);
  const projectContext = `Project Type: ${projectType}`;

  const prompt = template
    .replace("{{EQUIPMENT_DATA}}", equipmentData)
    .replace("{{PROJECT_CONTEXT}}", projectContext);

  const providersToTry: LLMProvider[] = [];
  if (isProviderAvailable(preferredModel)) {
    providersToTry.push(preferredModel);
  }
  const allProviders = getAvailableProviders();
  for (const p of allProviders) {
    if (!providersToTry.includes(p)) {
      providersToTry.push(p);
    }
  }

  if (providersToTry.length === 0) {
    throw new Error("No LLM providers available for vendor list generation");
  }

  let lastError: Error | null = null;
  for (const provider of providersToTry) {
    try {
      console.log(`Vendor List: Trying provider ${provider}...`);
      const result = await llmComplete({
        model: provider,
        messages: [
          { role: "system", content: prompt },
          { role: "user", content: "Generate the recommended vendor list for the equipment listed above. Return only valid JSON." },
        ],
        maxTokens: 8192,
        jsonMode: true,
      });

      let parsed: any;
      try {
        let content = result.content.trim();
        const fenceMatch = content.match(/```(?:json)?\s*([\s\S]*?)```/);
        if (fenceMatch) content = fenceMatch[1].trim();
        parsed = JSON.parse(content);
      } catch (parseErr) {
        throw new Error(`Failed to parse vendor list JSON: ${(parseErr as Error).message}`);
      }

      if (!parsed.items || !Array.isArray(parsed.items)) {
        throw new Error("Vendor list response missing 'items' array");
      }

      for (const item of parsed.items) {
        if (!item.equipmentId || !item.equipmentType) {
          throw new Error("Vendor list item missing required fields (equipmentId, equipmentType)");
        }
        if (!Array.isArray(item.recommendations)) {
          item.recommendations = [];
        }
        item.quantity = item.quantity || 1;
        item.specsSummary = item.specsSummary || "";
        item.process = item.process || "";
      }

      const vendorList: VendorList = {
        items: parsed.items,
        generatedAt: new Date().toISOString(),
        modelUsed: providerLabels[provider] || provider,
      };

      return { vendorList, providerLabel: providerLabels[provider] || provider };
    } catch (err) {
      lastError = err as Error;
      console.warn(`Vendor List: Provider ${provider} failed:`, lastError.message);
    }
  }

  throw lastError || new Error("All providers failed for vendor list generation");
}
